Apps/DE/EDocumentFormatXRechnung/app/src/ExportXRechnungDocument.Codeunit.al

```diff
@@ -0,0 +1,1117 @@
+        SalesInvLine.SetFilter(
+          "VAT Calculation Type", '%1|%2|%3',
+          SalesInvLine."VAT Calculation Type"::"Normal VAT",
+          SalesInvLine."VAT Calculation Type"::"Full VAT",
+          SalesInvLine."VAT Calculation Type"::"Reverse Charge VAT");
+        SalesInvLine.CalcSums(Amount, "Amount Including VAT");
+        SalesInvLine.SetRange("VAT Calculation Type");
+        exit(SalesInvLine."Amount Including VAT" - SalesInvLine.Amount);
+    end;
+
+    local procedure GetTotalTaxAmount(var SalesCrMemoLine: Record "Sales Cr.Memo Line"): Decimal
+    begin
+        SalesCrMemoLine.SetFilter(
+          "VAT Calculation Type", '%1|%2|%3',
+          SalesCrMemoLine."VAT Calculation Type"::"Normal VAT",
+          SalesCrMemoLine."VAT Calculation Type"::"Full VAT",
+          SalesCrMemoLine."VAT Calculation Type"::"Reverse Charge VAT");
+        SalesCrMemoLine.CalcSums(Amount, "Amount Including VAT");
+        SalesCrMemoLine.SetRange("VAT Calculation Type");
+        exit(SalesCrMemoLine."Amount Including VAT" - SalesCrMemoLine.Amount);
+    end;
+
+    local procedure GetCurrencyCode(DocumentCurrencyCode: Code[10]; var Currency: Record Currency): Code[10]
+    begin
+        if DocumentCurrencyCode = '' then begin
+            Currency.InitRoundingPrecision();
+            exit(GeneralLedgerSetup."LCY Code");
+        end else begin
+            Currency.Get(DocumentCurrencyCode);
+            Currency.TestField("Amount Rounding Precision");
+            Currency.TestField("Unit-Amount Rounding Precision");
+            exit(DocumentCurrencyCode);
+        end;
+    end;
+
+    local procedure GetCountryRegionCode(CountryRegionCode: Code[10]): Code[10]
+    begin
+        if CountryRegionCode <> '' then
+            exit(CountryRegionCode);
+
+        CompanyInformation.TestField("Country/Region Code");
+        exit(CompanyInformation."Country/Region Code");
+    end;
+
+    procedure GetVATRegistrationNo(VATRegistrationNo: Text[20]; CountryRegionCode: Code[10]): Text[30];
+    begin
+        if CopyStr(VATRegistrationNo, 1, 2) <> CountryRegionCode then
+            exit(CountryRegionCode + VATRegistrationNo);
+        exit(VATRegistrationNo);
+    end;
+
+    local procedure DocumentLinesExist(SalesInvoiceHeader: Record "Sales Invoice Header"; var SalesInvLine: Record "Sales Invoice Line"): Boolean
+    begin
+        SalesInvLine.SetRange("Document No.", SalesInvoiceHeader."No.");
+        SalesInvLine.SetFilter(Type, '>%1', 0);
+        SalesInvLine.SetFilter("No.", '<>%1', '');
+        SalesInvLine.SetFilter(Quantity, '<>0');
+        if SalesInvLine.FindSet() then
+            exit(true);
+        exit(false);
+    end;
+
+    local procedure DocumentLinesExist(SalesCrMemoHeader: Record "Sales Cr.Memo Header"; var SalesCrMemoLine: Record "Sales Cr.Memo Line"): Boolean
+    begin
+        SalesCrMemoLine.SetRange("Document No.", SalesCrMemoHeader."No.");
+        SalesCrMemoLine.SetFilter(Type, '>%1', 0);
```

Comment:
Those filters here aren't the best practice.

We shouldn't use integer values for Enums.
Therefore I would suggest to use something like 
`SalesCrMemoLine.SetFilter(Type, '<>%1', SalesCrMemoLine.Type::" ");` instead of `SalesCrMemoLine.SetFilter(Type, '>%1', 0);`.


Furthemore the filter for an empty `No. `field is wrong.
it should be `SalesCrMemoLine.SetFilter("No.", '<>%1', '');` instead of `SalesCrMemoLine.SetFilter("No.", '<>%1', ' ');`
(I removed the space in the filter):
Have you planned to add there any events?

Reply:
apart from using enum values and not integers, please also use the SetFilter syntax consistently. This code uses both the 2 and 3 param overloads inconsistently. 
If you use the 2 param syntax, you get no type checking. It is preferred to use the 3 param one for that reason, so:
SetFilter(Quantity, '<>%1', 0) will check that 0 matches the type of Quantity...

ex:
SetFilter(Quantity, '<>O'); => this would compile
SetFilter(Quantity, '<>%1', O); => this would not

Reply:
This is a sync between our internal repo and GitHub, and not a developer PS as such. That of course doesn't mean that the comments aren't valid, but it goes mean that the code already is checked in. I will create a bug for these comments.